### ISSUE TITLE
SC2016: Slightly too eager?

### ISSUE THREAD

**Issue by mietek** [koalaman/shellcheck#279]
```
In setup.sh line 56:
echo 'source <( "${HOME}/halcyon/halcyon" paths )' >>"${HOME}/.bash_profile"
     ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.
```

I don’t think ShellCheck should be making a suggestion in this case. I am using single quotes _because_ I don’t want this expression to be expanded.

**Comment by kballard** [comment]
I too am having this issue. I use single-quotes specifically to disable expansion. If I wasn't using characters that triggered expansion in double-quoted strings then I would have just used a double-quoted string to begin with. In my case, I have code like

```sh
echo 'expected parameter to `fatal` flag -n'
```

I'm actually having a hard time figuring out where this warning would be useful to anyone writing scripts with a syntax-aware editor. In fact, this particular warning suggests that perhaps ShellCheck needs a variant of the directive that applies to the entire script.

**Comment by koalaman** [comment]
@mietek It shouldn't, but how can shellcheck differentiate this from things like people do all the time like `echo '<img src="$IMAGE">' >> public_html/index.html` ? 

@kballard Believe it or not, this warning is actually one of the most frequently useful ones. Unfortunately, once you have a basic grasp of quoting, it becomes completely useless. 

When you're aware of the difference, I'd suggest just using `-e 2016` always, and there should probably be support for a `.shellcheckrc` where you can add these issues once you outgrow them, so to speak.

**Comment by kballard** [comment]
@koalaman A `.shellcheckrc` would be nice, but I think there should be a shellcheck directive I can put at the very top of the file that applies to the rest of the file. That way other people who use ShellCheck won't see warnings like SC2016 that I might have disabled in a `.shellcheckrc`.

**Comment by koalaman** [comment]
@kballard That's a good point.

How do you feel about using a top level brace group? It lets you do this right now and doesn't incur an overhead during execution:

    #!/bin/bash
    #shellcheck disable=SC2016
    {
    echo '$foo'
    echo '$bar'
    }

**Comment by kballard** [comment]
@koalaman That's problematic. Not only will my editor then want to indent the entire file, but Vim's syntax coloring apparently doesn't color the contents of a brace list the same way it does the top level, meaning things like function definitions won't be syntax-highlighted correctly.

**Comment by kballard** [comment]
Suggestion: support a directive of the form

```sh
#shellcheck global disable=SC2016
```

This could either affect the rest of the file, or it could be required to come before any non-empty non-commented lines.

**Comment by mietek** [comment]
@koalaman: This issue isn’t very important to me, as I can very easily work around it by doing some manual quoting. However, I do think that this diagnostic is borderline useless — as you say yourself, the difference between single and double quotes is Shell Scripting 101. Moreover, quoting something by mistake should be an easily-caught error — if you test the script, you will immediately see the problem.

**Comment by koalaman** [comment]
A `grep sed | grep variable` style search on the freenode #bash log shows that it's not always immediate and obvious to everyone (this is obviously just for sed and just for people who realize it's a problem with expansion):

    < xxxx> How can I get sed to read variables? like echo $i | sed 's/$PWD//' except that doesn't work because it doesn't read what's in the variable.
    < xxxx> cat file.xml | sed 's/PATTERN/$BASH_VARIABLE/g' does not resolve the variable
    < xxxx>  how can I have sed/bash escape chars inside a variable so I could do something like sed -e 's/$VARIABLE//g' ?
    < xxxx> greycat: you know of a way to do something similar to this concept:  echo "replace THIS word"  | sed 's/THIS/$blah'   and have $blah expanded before it replaces THIS
    < xxxx> How can I use sed to substitute a word with a variable? Like; sed 's/^/$DaTe   /g'
    < xxxx> how do I use variables within sed? i.e. sed -e 's/$VAR//'
    < xxxx> what should i be using in place of the following, for it to actually expand the $versionnumbers variable when doing the substitution? "sed -i 's/%VERSIONNUMBERS%/$versionnumbers/' template/README.linux"
    < xxxx> koala_man, if i would then add a word--> $word , at the end  sed -i '1  s/[^/]*$/$word/' .. this doesn't consider $word as a variable
    < xxxx> i want to substitute this line `echo $1 | cut -d . -f 2` with this line "$file_extension" in a file named audio-convert. what's wrong with this command? sed 's/\`echo \$1 \| cut \-d \. \-f 2\`/\"\$file_extension\"/g' audio-convert. help! : ) thankyou! : )
    < xxxx> quick bash/sed question -- If i want to use variables $A and $B in a sed function, (  sed 's/$A/$B/g' ) within a bash script, How can I get those variables to expand?  I know the single quotes prevent that.
    < xxxx> sed '/$name/s/foo/bar/' file_name
    < xxxx> r00t|vaio, np :-) i'm trying,  sed -e 's/(.*)(\$row.*)\=(.*)\;/$3=$2/g' < foo.txt
    < xxxx> I use sed to reemplace a existing variable but with no result: var=foo; sed -i 's/$var/bar/g' examplefile
    < xxxx> hi all, quick question, im trying to use sed (find&replace) such as: sed -i 's/text/$VAR/g' file , but it just replaces text=$VAR and not the contents of the var, how can I do this?
    < xxxx> can I get a bash script to interpret a variable inside single quotes? sed -e 's/something/$MY_VARIABLE/g' filename
    < xxxx> ummm... how do i use sed -i to replace a directory path as in sed -i 's//home/user//$variable/g' /file
    < xxxx> Something like: sed -i '1s/^.*$/&\n$(md5sum...)'/' bar  ?
    < xxxx> yes, i want to do something like sed 's/foo/foo$v' file
    < xxxx> ok, screen takes in a config file:  screen -c /path/to/file -S foo.  my /path/to/file has some variables that need to be replaced, but not permanently.  so how can i feed screen something like:  screen -c sed 's/$VAR1/hi/g' /path/to/file -S foo
    < xxxx> guys, i am trying to replace one variable by another in a file using sed like this, cat thesis.tst1 |sed -e 's/$fld1/$fld2/g' >bar
    < xxxx> How can I substitute a line with the content of a variable with sed? sed -i 's/str/$VAR/g' file doesn't work, I think it's because $ is a special character used for regex... So, how should I do that?
    < xxxx> dooglus: can I insert the contents of a variable into a file using sed -i '1 i $VARIABLE'  <filenane>?
    < xxxx>  is it possible expand a variable into sed? i.e.: sed 's/$foo/bar/' file.txt - $foo = letter
    < xxxx> ok for bash is it possible to make the pattern a variable ? like pattern=Dynamic Network Services  then it would be like sed -i '$pattern' file.txt ..i just couldnt get it working
    < xxxx> i have sed -i '3s/.*/DEFAULT_PROXY=$DEFAULT_PROXY/' ./file . Why isn't $DEFAULT_PROXY replaced by its correct value?
    < xxxx> Hello. Is it possible to use a variable in a sed replacing line? as in, sed 's/a \(foo\) b/a $VAR b' ?
    < xxxx> TheBonsai: can I use also bash-variables when using for example sed? (so sed -e 's/^/$BASHVARIABLE1;$BASHVARIABLE2;/' file)
    < xxxx> sed 's/:/$j\:/g' or similar i dont thinkthe variable will expand
    < xxxx> damn i still have one more problem :( sed -i 's/1[0-9]\{9\}/$TIMESTAMP/g' $ZONEFILE <- i want it to print the content of the variable, instead it just writes "$TIMESTAMP"... how can i make it behave like i want?
    < xxxx> what's wrong here? how can I get my variable to expand inside sed? name=foo; echo \"\" | sed 's/""/$name/'

Any solution would have to consider both camps.

**Comment by mietek** [comment]
Similarly, SC2088 always warns about a quoted tilde (`~`), regardless of the quote type:
```
In src/buildpack.sh line 121:
		log_indent "~ $ restore"
                            ^-- SC2088: Note that ~ does not expand in quotes.
```

**Comment by koalaman** [comment]
Well that's just annoying. With 0a2cf20 it will only trigger when followed by a slash (no one uses `~user/` syntax anyways)  and `\~` will not claim to become a regular tidle.

**Comment by koalaman** [comment]
Directives can now be put on the top of the file to apply to everything, and there's also `SHELLCHECK_OPTS` for implicitly ignoring all errors of some type. Closing this issue.
